### PR TITLE
fix(Scripts/Spell): Volley should trigger pet attack in defensive mode.

### LINKED ISSUE #7340: Hunter Volley Not Activating Pet Attack in Defensive Mode
### What client do you play on?

enUS

### Faction

- [ ] Alliance
- [X] Horde

### Content Phase:

- [X] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59

### Current Behaviour
Original report: https://github.com/chromiecraft/chromiecraft/issues/1408

does not provoke a pet to attack (pet in defence mode):
WoWScrnShot_081421_103718
all other attacks lead to pet attack. In instances\dungeons pet just stay when i use [Volley]

Tester: confirmed. All other hunter abilities 9that I tried at least) will still make a pet in Defensive mode attack your target. However, Volley doesn't.

https://user-images.githubusercontent.com/81782124/129450407-c7050e74-2c9b-466a-af12-31fb55facfea.mp4

### Expected Blizzlike Behaviour

attack

### Source

"...a defensive pet will immediately attack any mob which the hunter attacks." - https://wowwiki-archive.fandom.com/wiki/Defensive?oldid=2166398

There's an argument from consistency - there's nothing in Volley's description or comments on TBC/Wowhead (see https://wowpedia.fandom.com/wiki/Volley?oldid=2358743) to indicate that a pet might behave differently while it was active.

Also some minor evidence from Petopia:
"since I'm spamming Volley by that point I've hit every mob at least once, so my Defensive pets will always redirect to the next mob when the one they're targeting is dead." - https://forums.wow-petopia.com/viewtopic.php?t=1961

### Steps to reproduce the problem

have pet in defence mode
use [Volley]

### Extra Notes

_No response_

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/2d77fe9274005a99f8445b8dac9dd0fcd3faf3c5

### Operating system

Ubuntu 20.04

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-CarbonCopy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-CustomWorldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-LevelUpReward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-Zonecheck](https://github.com/55Honey/acore_Zonecheck)
- [lua-TempAnnouncements](https://github.com/55Honey/Acore_TempAnnouncements)

### Customizations

None

### Server

ChromieCraft

### PR DESCRIPTION
**Pull request by UltraNix**
Fixed #7340

## Changes Proposed:
-  Volley should trigger pet attack in defensive mode.

## Issues Addressed:

- Closes #7340
- Closes https://github.com/chromiecraft/chromiecraft/issues/1408

## Tests Performed:

- Tested ingame.

## How to Test the Changes:

1. have pet in defence mode
2. use [Volley] and watch your pet attack the target.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/scripts/Spells/spell_hunter.cpp [hunk 1]
```diff
@@ -1229,6 +1229,63 @@ class spell_hun_viper_attack_speed : public SpellScriptLoader
...
+            // Let's select one of our targets
+            if (!targets.empty())
+            {
+                _target = *(targets.begin());
+            }
+        }
+
+        void HandleFinish()
+        {
+            if (!_target || _target->GetTypeId() != TYPEID_UNIT)
```
Comment (lineagedr) [inline src/server/scripts/Spells/spell_hunter.cpp:1254]:
```c++
_target->GetTypeId() != TYPEID_UNIT
```
What if the target is a player? Then the pet won't attack on volley.

Comment (UltraNix) [reply src/server/scripts/Spells/spell_hunter.cpp:1254]: Yep - you are right :)

File: src/server/scripts/Spells/spell_hunter.cpp [hunk 2]
```diff
@@ -1229,6 +1229,63 @@ class spell_hun_viper_attack_speed : public SpellScriptLoader
...
+        }
+
+        void HandleFinish()
+        {
+            if (!_target)
+                return;
+
+            Unit* caster = GetCaster();
+            if (!caster || !caster->IsPlayer())
+                return;
```
Comment (Kitzunu) [inline src/server/scripts/Spells/spell_hunter.cpp:1259]: bracket

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by pklloveyou** [comment]
Tested

**Comment by Kitzunu** [comment]
Conflicts

**Pull request merged**